ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/truncate/TruncateTableAnalyzer.java

```diff
@@ -338,4 +373,42 @@ private void addStatTask(ASTNode root, Table table, Path oldPartitionLocation, P
       moveTask.addDependentTask(statTask);
     }
   }
+
+  public StringBuilder constructDeleteQuery(Table table, Map<String, String> partitionSpec) throws SemanticException {
+    StringBuilder sb = new StringBuilder().append("delete from ").append(table.getTableName())
+            .append(" where ");
+    List<String> keyList = new ArrayList<String>(partitionSpec.keySet());
+    Deserializer deserializer = table.getDeserializer();
+    Map<String, PrimitiveObjectInspector.PrimitiveCategory> stringTypeInfoMap = new HashMap<>();
+    try {
+      ObjectInspector objectInspector = deserializer.getObjectInspector();
+      if (objectInspector.getCategory() == ObjectInspector.Category.STRUCT) {
+        StructObjectInspector structObjectInspector = (StructObjectInspector) objectInspector;
+        List<? extends StructField> structFields =  structObjectInspector.getAllStructFieldRefs();
+        for (int index = 0;index < structFields.size();index++) {
+          StructField structField = structFields.get(index);
```

Comment:
nit.: Is it possible to use foreach here?
```
for (StructField structField : structFields) {
...
```

Reply:
Done